Chapter03/Activity3.01/app/src/main/java/com/example/planetquiz/AnswersFragment.kt

```diff
@@ -4,13 +4,21 @@ import android.os.Bundle
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import android.widget.TextView
 import androidx.fragment.app.Fragment
-import kotlinx.android.synthetic.main.fragment_answers.*
 
 class AnswersFragment : Fragment(), View.OnClickListener {
 
     var questionId: Int = NO_QUESTION_SET
 
+    private val headerText by lazy {
```

Comment:
This will lead to bugs. When the view is recreated, this will hold a dead reference. Use get() instead.

Reply:
When the view will be recreated the private val is assigned again, I am not sure how will it lead to bugs and where will be the dead reference?

Reply:
It will not be reassigned. The gragment is retained, the view re-inflated. These references will not be re-evaluated.

Reply:
I had a play around with this, just logging the hashcode of the fragment on rotation, so recreating the activity and the fragment was recreated. Won't the fragment only be retained if setRetainInstance(true) is set?

Reply:
No. A Fragment may persist and its view may not. This isn't about rotation. Of course, in that case the Fragment itself is destroyed.

Take a look here:
https://www.google.com/amp/s/mohamedmoanes.wordpress.com/2018/07/29/fragment-lifecycle/amp/

Reply:
@drulabs - Apologies I was working on chapter 3 code updates and messed up the merge of my changes into your branch. It hasn't messed up the history thankfully. All your code has been merged @drulabs. You can view the commit here: [Squashed commit with changes](https://github.com/PacktPublishing/How-to-Build-Android-Apps-with-Kotlin/commit/67cc1132208421a436510c1c8a9e15b6ede9cce8 Obviously the issue with the findViewById best approach exists.)

Thanks for the article @EranBoudjnah - I realise fragments can be created and then added and removed from containers etc. whilst the fragment is still retained without a view. In the scenario above I think that when the layout is created and available after onCreateView has finished then properties like headerText above can use the view objects. Interestingly when a layout has been inflated in onCreateView, but not returned from the function then accessing the field will be null as would be expected. Once the view is returned from onCreateView then the property reference of headerText above will correctly be the header_text view. I appreciate if you weren't attaching a view then it wouldn't be. I don't personally think the syntax above is likely to cause issues as we are not using headless fragments although the syntax for a getter:
` private val headerText get() = view?.findViewById<TextView>(R.id.header_text)` could also be used. The issue is consistency really. In all the previous examples in Fragment classes @drulabs has used 'by lazy' so if this change is to be made it would need to be made in Chapters 3 and possibly other chapters in the book. The issue here is that property getters and setters would have to be explained for their first use so chapter 3 would need this explanation and also all code examples in the chapter and on github being updated. So this will obviously take more time. So wondering whether we could leave this as it is interested to hear your thoughts @EranBoudjnah @drulabs

Reply:
@alexandrudumbravan let me start by saying I totally get your frustration at having to revisit this _yet again_.

The thing is, when you develop a class, you don't know how it will be used.  The same is true for fragments. It is not unlikely they will be detached and reattached and so a view re-inflated.

When that happens, nothing you do with the lazy loaded views will work: you won't be able to set their visibility, you won't be able to respond to clicks on them, you won't be able to set text on them or load images into them. This will seem awfully odd, because when debugging you will see you have a valid instance.

What I'm trying to say is this is setting up the reader for failure by teaching them this bad practice.

We can solve this in one of two ways: either we use `get()` as suggested above, instead of lazy. Given the reader is expected to have knowledge of Kotlin, we shouldn't have to explain this in great depth.

Or, we drop the field altogether and use `findViewById()` where needed.

It's worth noting this is only true for fragments. Activities and custom views are fine to use lazy.

Also worth noting is you can use `requireView()` if opting for `get()`.

Reply:
Hey @EranBoudjnah - I think you tagged the wrong Alex! :-) I'm @alexforrester - That's fair enough about the concerns about using 'by lazy'. When I tested attaching, detaching and then re-attaching the same fragment it did create a new instance of the fragment. Regardless your concerns about Fragment and View lifecycles are valid so it will be good to change it.

@drulabs Could you update the github code and chapter examples using get() as in?
`private val headerText get() = view?.findViewById<TextView>(R.id.header_text)` as an example
There will also have to be a small explanation of getter/setter property use in chapter 3 the first time this is used - I appreciate this is an additional change you'll have to make so thank you very much for this.

As these properties are used in multiple places in individual fragments I think it's preferable to do this rather than use findViewById. This will be required for all properties in Fragment classes in Chapter 3 - thank you

@alexandrudumbravan @jomartigcal - Can you confirm that there are no fragment properties using 'by lazy' in your chapters? If there are then they will have to change to use the get() syntax - thank you

Reply:
@EranBoudjnah - sorry I misspoke there. It attached the same fragment with a new view